drivers/i2s/i2s_mcux_flexcomm.c

```diff
@@ -0,0 +1,940 @@
+/* This function is executed in the interrupt context */
+static void i2s_mcux_dma_tx_callback(const struct device *dma_dev, void *arg,
+				uint32_t channel, int status)
+{
+	const struct device *dev = (struct device *)arg;
+	struct i2s_mcux_data *dev_data = dev->data;
+	struct stream *stream = &dev_data->tx;
+	void *buffer;
+	int ret;
+
+	LOG_DBG("tx cb: %d", stream->state);
+	ret = k_msgq_get(&stream->out_queue, &buffer, K_NO_WAIT);
+	if (ret == 0) {
+		/* transmission complete. free the buffer */
+		k_mem_slab_free(stream->cfg.mem_slab, &buffer);
+	} else {
+		LOG_ERR("no buffer in output queue for channel %u", channel);
+	}
+
+	/* Received a STOP trigger, terminate TX immediately */
+	if (stream->last_block) {
+		stream->state = I2S_STATE_READY;
+		i2s_mcux_tx_stream_disable(dev, false);
+		LOG_DBG("TX STOPPED");
+		return;
+	}
+
+	switch (stream->state) {
+	case I2S_STATE_RUNNING:
+	case I2S_STATE_STOPPING:
+		/* get the next buffer from queue */
+		ret = k_msgq_get(&stream->in_queue, &buffer, K_NO_WAIT);
+		if (ret == 0) {
+			/* config the DMA */
+			i2s_mcux_config_dma_blocks(dev, I2S_DIR_TX, (uint32_t *)buffer);
+			k_msgq_put(&stream->out_queue, &buffer, K_NO_WAIT);
+			dma_start(stream->dev_dma, stream->channel);
+		}
+
+		if (ret || status) {
+			/*
+			 * DMA encountered an error (status != 0)
+			 * or
+			 * No bufers in input queue
+			 */
+			LOG_DBG("DMA status %08x channel %u k_msgq_get ret %d",
+				status, channel, ret);
+			if (stream->state == I2S_STATE_STOPPING) {
+				stream->state = I2S_STATE_READY;
+			} else {
+				stream->state = I2S_STATE_ERROR;
+			}
+			i2s_mcux_tx_stream_disable(dev, false);
+		}
+		break;
+	case I2S_STATE_ERROR:
+		i2s_mcux_tx_stream_disable(dev, true);
+		break;
+	}
+}
+
+static void i2s_mcux_dma_rx_callback(const struct device *dma_dev, void *arg,
+				uint32_t channel, int status)
+{
+	struct device *dev = (struct device *)arg;
+	struct i2s_mcux_data *dev_data = dev->data;
+	struct stream *stream = &dev_data->rx;
+	void *buffer;
+	int ret;
+
+	LOG_DBG("rx cb: %d", stream->state);
+
+	if (status != 0) {
+		stream->state = I2S_STATE_ERROR;
+		i2s_mcux_rx_stream_disable(dev, false);
+		return;
+	}
+
+	switch (stream->state) {
+	case I2S_STATE_STOPPING:
+	case I2S_STATE_RUNNING:
+		/* retrieve buffer from input queue */
+		ret = k_msgq_get(&stream->in_queue, &buffer, K_NO_WAIT);
+		assert(ret == 0);
+
+		/* put buffer to output queue */
+		ret = k_msgq_put(&stream->out_queue, &buffer, K_NO_WAIT);
+		if (ret != 0) {
+			LOG_ERR("buffer %p -> out_queue %p err %d", buffer,
+				&stream->out_queue, ret);
+			i2s_mcux_rx_stream_disable(dev, false);
+			stream->state = I2S_STATE_ERROR;
+		}
+		if (stream->state == I2S_STATE_RUNNING) {
+			/* allocate new buffer for next audio frame */
+			ret = k_mem_slab_alloc(stream->cfg.mem_slab, &buffer, K_NO_WAIT);
+			if (ret != 0) {
+				if (k_msgq_num_used_get(&stream->in_queue) > 0) {
+					LOG_DBG("wait for data buffer in queue to fill");
+					break;
+				}
```

Comment:
Now that we enforce that the DMA engine has always 2 descriptors I think we should remove this code.

Reply:
This is required as alloc could return an error if the slab has only 2 blocks and both blocks are allocated during START. We wait for the second block that is in the input queue to be also filled. We return an error after all blocks have been filled with RX data.

Reply:
This still looks problematic. The DMA engine is set up to work with 2 descriptors. We've taken one block (being pointed to by one of the descriptors) but don't put any replacement. So from this point on the DMA engine is going to work with one descriptor only, until we stop the RX stream.
- I was under impression the DMA engine doesn't work reliably with 1 descriptor only. It needs to have at least 2, so when one of the blocks is filled in it can immediately start filling in another, before the first one is read by this callback.
- We would need to test if the DMA driver works at all if it is configured with 2 DMA descriptors but only 1 is available.

Reply:
This has been tested with 2 buffers as the i2s_loopback test allocates 2 RX and TX buffers. Unfortunately in the 2 RX Buffer case as you pointed it would NOT be using both slots all the time unless there is a way to queue the buffer as soon as the user frees a memory block.

Reply:
Since we set up DMA engine to work with 2 descriptors it feels like a bad idea to loose 1 descriptor sometime in the middle of the transfer. If `if (ret != 0) {` branch is ever taken from that point on the transfer will continue with only 1 descriptor available to the DMA engine, for the rest of the transfer until user/application sends a request to stop. It's a weakness we probably shouldn't introduce.

It's OK to update the testcases to allocate 3 (or 4) RX buffers. 

In fact, looking at this, it seems like the driver was always running with 1 descriptor only. Application keeps hold of at least 1 buffer, so with 2 buffers total in the pool only 1 is available to the driver. That's probably the reason for the buffer underflow error you've observed in `test_i2s_state_stopping_neg` testcase.

Reply:
I have updated the implementation to check if can add queue additional buffers in the RX callback. This does not guarantee it will succeed as the user may still be holding on the buffers, however this code provides an opportunity to queue the second buffer.

Reply:
I don't think we should be trying to add to the queue additional buffers in the RX callback, unless I misunderstood what you mean. Let's just make sure we don't lose any. When we start the transfer the DMA engine holds ownership of 2 buffers, if `if (k_msgq_num_used_get(&stream->in_queue) > 0) {` branch is taken it loses ownership of one buffer and will never get it back. Or otherwise: we've just sent one buffer to the application, now we need to get a replacement for it. Let's remove this if branch.

It's OK to force application to prepare enough buffers so that 2 can be always owned by the DMA engine.

Have you tried to run my version of the `test_i2s_state_stopping_neg` testcase when the testcase is changed to allocate 3 or 4 RX buffers in the memory slab?

Reply:
The code I added increases the number of buffers in the RX queue to 2 if it was short. This may not have an impact in the 2 buffers scenario. 
Yes, I did try the your version and I was able to receive the data without any RX overflows. However the original issue persisted.

Reply:
Now, after removing the `if` branch, the code looks good.